src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/SafeCryptoHandles.cs

```diff
@@ -182,9 +178,34 @@ internal bool PublicOnly
             }
         }
 
+        internal void SetParent(SafeProvHandle parent)
+        {
+            if (IsInvalid || IsClosed)
+            {
+                return;
+            }
+
+            Debug.Assert(_parent == null);
+            Debug.Assert(!parent.IsClosed);
+            Debug.Assert(!parent.IsInvalid);
+
+            _parent = parent;
+
+            bool ignored = false;
+            parent.DangerousAddRef(ref ignored);
```

Comment:
If you have:
```C#
_parent = parent;
parent.DangerousAddRef(ref ignored);
```
then if `this` is finalized between those two lines, _parent.DangerousReleaseRef will be called with a 0 ref count, causing a failure.

If you instead have:
```C#
_parent = parent;
_parent.DangerousAddRef(ref ignored);
```
then `this` can't be finalized until after the addref because `this` is still rooted.

I don't understand the question about parent.DangerousAddRef in your "Options 2"... can you clarify?  Whether you do ```_parent.DangerousAddRef``` or ```parent.DangerousAddRef``` is immaterial with regards to the state of ```parent```.